Add sound feedback to most tab control implementations from keyboard interactions

frenzibyte opened the pull request:
- Closes #18551 
- Closes #27400
- Supersedes #27448 
- [x] Depends on https://github.com/ppy/osu-framework/pull/6218
- [x] Depends on framework package

Removes usage of `HoverClickSound` and replaces it with `HoverSounds` on all tab item implementations, and overrides the newly added event `OnUserTabSelectionChanged` to play sound instead, so that audio feedback gets extended to keyboard interactions such as Ctrl-Tab or by the `SelectItem` method (used in this PR to resolve #27400). 

I will say I'm not 100% sure about the direction of this PR after reading the final diff, the fact that the logic for audio feedback is now split across `TabControl` (click) and `TabItem` (hover) feels really bad from a code quality perspective, but I don't really have any brighter idea right now.

The weirdest part of this diff would be `HoverClickSounds.PlayAsClick`, but it's necessary to share the click playback logic between `HoverClickSounds` and `TabControl`. I can remove it and let `TabControl` play the sample normally without any frequency adjustment, if that's better.

frenzibyte committed:
Update `TabControl` usage in line with framework changes

frenzibyte committed:
Play tab selection sample using new user triggered event rather than `HoverClickSounds`

frenzibyte committed:
Use `SelectItem` when switching editor screen mode to play sample

frenzibyte committed:
Add sound feedback to tab selection change in `OsuTabControl`

frenzibyte committed:
Refactor certain `TabControl` implementations to support new path of sound feedback

frenzibyte committed:
Change to `HoverSounds` in existing tab item implementations

Joehuu reviewed (commented):
Apart from playlist/multiplayer currently crashing due to no `AddEnumEntriesAutomatically` override, there is no more click sound when pressing the already selected item. Just pointing that out and maybe change to the disabled sound if its worth fixing. Otherwise changes look good to me.

frenzibyte commented:
@peppy thoughts? Should a sound still be played when clicking on an already selected tab, or not play anything. I've been going with the latter until I had realised master does the former.

Also if a sound is to be played, should it be the normal click sound, or the disabled version?

peppy commented:
If it's a select sound it should only play when selection changes.

frenzibyte commented:
It is, there's no need for extra changes then. I'll resolve the multiplayer/playlists issues when framework is bumped here.

peppy committed:
Merge branch 'master' into add-sound-to-tab-selection-change

peppy committed:
Update framework

frenzibyte committed:
Fix online-play tab controls not updated correctly

frenzibyte committed:
Fix multiple-selection beatmap search tab items no longer playing click sound

frenzibyte committed:
Fix overlay ruleset selector not inheriting from `OsuTabControl`

`ToolbarRulesetSelector` has a completely different way of playing sounds, so I'm intentionally not making it inherit `OsuTabControl` (it's practically a custom implementation of a tab control).

frenzibyte committed:
Work around NRT weirdness

frenzibyte committed:
Fix beatmap ruleset selector incorrectly using `SelectTab` to set initial ruleset

frenzibyte committed:
Fix beatmap listing sort tab control not playing sound when changing sort direction

frenzibyte committed:
Fix overlay stream tab controls having spacing applied

frenzibyte committed:
Rename and reorder screen switcher field

frenzibyte commented:
I've fixed tests and resolved a few bugs with the implementation after checking previous diff. Should be in a good state for review.

peppy commented:
I have real questions about some of the fixes you made...

ie. why did https://github.com/ppy/osu/pull/27657/commits/57c12a220c291670d584f955cc1e60a82bcb98c0 stop working in the first place?

why are https://github.com/ppy/osu/pull/27657/commits/7ca7328e3696645bb4f81feced976fc5def53411 / https://github.com/ppy/osu/pull/27657/commits/a603a7ecb0249eac1c46b8aa981620cf280f805e in separate commits with separate messaging? also this handling feels pretty awkward..

Given the fixes are so wide-reaching, I'm not 100% sure how to review this other than click around the UI and check if things make sounds. Is that the best direction of review?

frenzibyte commented:
Main points to bring up:
 - I am moving as much `TabControl` implementations as possible to implement `OsuTabControl` for proper sound feedback on tab selection change (except for `RulesetSelector` because it plays its own sounds). I had moved the two classes in https://github.com/ppy/osu/commit/57c12a220c291670d584f955cc1e60a82bcb98c0 to inherit `OsuTabControl` but mistakingly did not update the implementations to work with [this logic in `OsuTabControl`](https://github.com/ppy/osu/blob/master/osu.Game/Graphics/UserInterface/OsuTabControl.cs#L72-L76), so I fixed that in 57c12a220c291670d584f955cc1e60a82bcb98c0.
 - With the new behaviour established in `OsuTabControl`s, using `SelectTab` has sound feedback, so in https://github.com/ppy/osu/commit/a603a7ecb0249eac1c46b8aa981620cf280f805e, I had changed that to use `Current.Value` because we don't want sound to be played in that particular scenario (executed when opening a beatmap info overlay and/or closing it).

https://github.com/ppy/osu/commit/7ca7328e3696645bb4f81feced976fc5def53411 is, as described, work around for NRT weirdness due to `TabControl` having NRT disabled while I have the affected classes (`ProfileRulesetSelector` and `BeatmapRulesetSelctor`) enabled. I can disable it back if preferred. https://github.com/ppy/osu/commit/a603a7ecb0249eac1c46b8aa981620cf280f805e also does the same (in addition to switching `SelectTab` with `Current.Value`), poor split of commits from my end.

peppy committed:
Remove silly `PlayAsClick` method

peppy committed:
Merge branch 'master' into add-sound-to-tab-selection-change

peppy commented on osu.Game/Overlays/BeatmapListing/BeatmapSearchMultipleSelectionFilterRow.cs:
```diff
@@ -118,6 +119,8 @@ public MultipleSelectionFilterTabItem(T value)
                 });
             }
 
+            protected override HoverSounds CreateHoverSounds() => new HoverClickSounds(HoverSampleSet.TabSelect);
```
I don't know how to explain why this is needed, but it **needs an inline comment**, or it needs to be removed.

frenzibyte replied:
Added inline comment in the alternative branch.

peppy commented on osu.Game/Overlays/OverlayPanelDisplayStyleControl.cs:
```diff
@@ -78,7 +78,7 @@ public PanelDisplayTabItem(OverlayPanelDisplayStyle value)
                         FillMode = FillMode.Fit
                     },
-                    new HoverClickSounds()
+                    new HoverSounds()
```
Why isn't the `TabSelect` style? And why can't it be shared in `OsuTabControl`.

frenzibyte replied:
That's how it was previously.

peppy commented:
I really hate to go through this many rounds of review for a PR like this. I'd hope it's something we could merge in 5 minutes because it's just putting sound effects in places. The reason it's hard is because we're trying to be smart and generic and reuse classes.

I wouldn't be opposed to just putting *every* instance of sound effects locally and calling it a day if that's seen as simpler. At least we wouldn't have to dance around with changes like this and waste literally hours of review time on it.

frenzibyte commented on osu.Game/Overlays/BeatmapListing/FilterTabItem.cs:
```diff
@@ -40,12 +40,14 @@ private void load()
             Enabled.Value = true;
         }
 
+        protected virtual HoverSounds CreateHoverSounds() => new HoverSounds(HoverSampleSet.TabSelect);
```
See above.

frenzibyte commented on osu.Game/Screens/OnlinePlay/Match/Components/MatchTypePicker.cs:
```diff
@@ -78,7 +80,7 @@ public GameTypePickerItem(MatchType value)
                         },
                     },
-                    new HoverClickSounds(),
+                    new HoverSounds(),
```
Went the opposite way around and brought back select sounds in my alternative branch, less loc.

frenzibyte commented on osu.Game/Screens/OnlinePlay/Match/Components/RoomAvailabilityPicker.cs:
```diff
@@ -75,6 +72,7 @@ public RoomAvailabilityPickerItem(RoomAvailability value)
                         Font = OsuFont.GetFont(weight: FontWeight.Bold),
                         Text = value.GetDescription(),
                     },
+                    new HoverSounds(),
```
See above.

frenzibyte commented:
The original intent of this PR was to move the triggering of the "tab select" sound completely from tab items and into `TabControl` itself to make it triggered by the `OnUserTabSelectChange` event (i.e. via click or via `SelectTab`). The downsides of that approach as it appeared in this PR are:
 1. Tab items still need to play hover sound, therefore `HoverSounds` still need to be implemented inside each `TabItem` implementation, making code readability quite weird due to sound not being handled in one place (i.e. https://github.com/ppy/osu/pull/27657#discussion_r1546048438)
 2. Every tab control implementation that we want to play select sound must inherit `OsuTabControl` and make sure their tab items use `HoverSounds` into `HoverClickSounds`, as to not cause double select sample playback.
 3. There are tab controls wherein their tab items have custom behaviour when clicking on them (e.g. beatmap listing "sort by" tab control).

If we go back all the way to https://github.com/ppy/osu/pull/27448, my intention there was to define an explicit method in `OsuTabControl` to play sound when selecting a tab, similar to what you're proposing by putting every instance of sound effect locally.

If you were to ask for my opinion, I'm thinking of the following as a last attempt: 
 - https://github.com/ppy/osu-framework/compare/master...frenzibyte:osu-framework:tab-control-user-events-2
 - https://github.com/ppy/osu/compare/master...frenzibyte:osu:add-sound-to-tab-selection-change-2

This branch aims to treat the downsides above by keeping "tab select" sound playback inside each tab item, avoiding the entire idea of making every tab control implement `OsuTabControl`, allowing custom tab items to still trigger select sound or completely roll off its own way of having it trigger (e.g. `OverlaySortTabControl<T>.SortTabItem`s), and greatly simplifying diff.

Notes:
 - The existence of `HoverSelectSounds` just serves as an easy way of being able to play select sample without having to manually look it up from the sample store on every tab item implementation. If it's very bad to exist, suggestions are welcomed.

peppy commented:
I don't get it

![2024-04-17 11 25 13@2x](https://github.com/ppy/osu/assets/191335/2ae54046-ad33-44c1-9231-cbab8352293f)

Why is clicking playing a select sound?

Why do we need

`HoverClickSounds : HoverSelectSounds : HoverSounds`

Too much inheritance. Too much nonsense.

frenzibyte commented:
This part that you are highlighting can be easily get ridden off by manually fetching the select sample in each tab item class, it does not play any role in the proposal.

peppy commented:
Okay let's give it a try then. It seems better than this PR at lesat.
